File: app/src/main/java/com/example/customprogressbar/remainingQuiteTime/models/RemainingQuiteTime.java
package com.example.customprogressbar.remainingQuiteTime.models;

import java.util.List;
import java.util.Locale;
import java.util.Objects;

public class RemainingQuiteTime {
    private static final int secondsInHours = 3600;
    private static final int secondsInMinutes = 60;

    private final List<QuiteTimeUser> quiteTimeUsers;
    private int secondsRemaining;

    public RemainingQuiteTime(List<QuiteTimeUser> quiteTimeUsers, int secondsRemaining) {
        this.quiteTimeUsers = quiteTimeUsers;
        this.secondsRemaining = secondsRemaining;
    }

    public RemainingQuiteTime(List<QuiteTimeUser> quiteTimeUsers) {
        this(quiteTimeUsers, 0);
    }

    public void decrementSecondsRemainig() {
        this.secondsRemaining--;
    }

    public int getSecondsRemaining() {
        return secondsRemaining;
    }

    public boolean isFinished() {
        return secondsRemaining == 0;
    }


    public String getFormattedTimeRemaining() {
        int hours = secondsRemaining / secondsInHours;
        int minutes = (secondsRemaining % secondsInHours) / secondsInMinutes;
        int seconds = secondsRemaining % secondsInMinutes;

        String formattedSeconds = String.format(Locale.ENGLISH, "%ds", seconds);
        String formattedMinutes = String.format(Locale.ENGLISH, "%dm", minutes);
        String formattedHours = String.format(Locale.ENGLISH, "%dh", hours);

        StringBuilder formattedTime = new StringBuilder();

        if(hours != 0 && minutes != 0) {
            formattedTime
                    .append(formattedHours)
                    .append(" ")
                    .append(formattedMinutes)
                    .append(" ");
        } else if(hours == 0 && minutes != 0) {
            formattedTime
                    .append(formattedMinutes)
                    .append(" ");
        }
        formattedTime.append(formattedSeconds);

        return formattedTime.toString();
    }


    public List<QuiteTimeUser> getQuiteTimeUsers() {
        return quiteTimeUsers;
    }
}

File: app/src/main/java/com/example/customprogressbar/remainingQuiteTime/models/RemainingQuiteTime.kt
package com.example.customprogressbar.remainingQuiteTime.models

import android.text.Spannable
import android.text.SpannableString
import android.text.style.RelativeSizeSpan
import java.lang.IndexOutOfBoundsException

class RemainingQuiteTime constructor(
        val quiteTimeUsers: ArrayList<QuiteTimeUser>,
        private var secondsRemaining: Int = 0) {

    companion object {
        private const val secondsInHours = 3600
        private const val secondsInMinutes = 60
    }

    var stopped: Boolean = false

    val isFinished: Boolean
        get() = secondsRemaining == 0


    val formattedTimeRemaining: SpannableString
        get() {
            val areDoubleDigits = arrayListOf<Boolean>()
            val hours = secondsRemaining / secondsInHours
            val minutes = secondsRemaining % secondsInHours / secondsInMinutes
            val seconds = secondsRemaining % secondsInMinutes

            var formattedTime = ""
            if (hours != 0 && minutes != 0) {
                formattedTime = "${hours}h ${minutes}m "
                areDoubleDigits.add(hours/10 > 0)
                areDoubleDigits.add(minutes/10 > 0)
            } else if (hours == 0 && minutes != 0) {
                formattedTime = "${minutes}m "
                areDoubleDigits.add(minutes/10 > 0)
            }
            formattedTime += "${seconds}s"
            areDoubleDigits.add(seconds/10 > 0)

            return createSpannableString(formattedTime, areDoubleDigits)
        }

    private fun createSpannableString(formattedTime: String, areDoubleDigits: List<Boolean>): SpannableString {
        val spannable = SpannableString(formattedTime)
        var currentIndex = 0
        for(isDoubleDigit in areDoubleDigits) {
            val numberSpace = if (isDoubleDigit) 2 else 1
            spannable.setRelativeSizeSpan(currentIndex, currentIndex + numberSpace)
            currentIndex += numberSpace + 2
        }
        return spannable
    }

    fun decrementSecondsRemainig() {
        this.secondsRemaining--
    }
}

fun SpannableString.setRelativeSizeSpan(startIndex: Int, endIndex: Int) {
    val biggerFontRatio = 1.67f

    try {
        setSpan(RelativeSizeSpan(biggerFontRatio), startIndex,
                endIndex, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
    } catch (ignore: IndexOutOfBoundsException){}
}